File: frontend/src/graphql/preview.ts
import { gql } from '@apollo/client';
import { Layer } from './graphs';

// DataSet Preview Query
export const GET_DATASOURCE_PREVIEW = gql`
  query GetDataSetPreview(
    $projectId: Int!
    $nodeId: String!
    $limit: Int
    $offset: Int
  ) {
    datasetPreview(
      projectId: $projectId
      nodeId: $nodeId
      limit: $limit
      offset: $offset
    ) {
      nodeId
      datasetId
      name
      filePath
      fileType
      totalRows
      columns {
        name
        dataType
        nullable
      }
      rows {
        rowNumber
        data
      }
      importDate
      executionState
      errorMessage
    }
  }
`;

// Graph Preview Query
export const GET_GRAPH_PREVIEW = gql`
  query GetGraphPreview($projectId: Int!, $nodeId: String!) {
    graphPreview(projectId: $projectId, nodeId: $nodeId) {
      nodeId
      graphId
      name
      annotations
      nodes {
        id
        label
        layer
        weight
        isPartition
        attrs
      }
      edges {
        id
        source
        target
        label
        layer
        weight
        attrs
      }
      layers {
        id
        layerId
        name
        backgroundColor
        textColor
        borderColor
        alias
        comment
        properties
      }
      nodeCount
      edgeCount
      executionState
      computedDate
      errorMessage
    }
  }
`;

// TypeScript Interfaces

export interface TableColumn {
  name: string;
  dataType: string;
  nullable: boolean;
}

export interface TableRow {
  rowNumber: number;
  data: Record<string, any>;
}

export interface DataSetPreview {
  nodeId: string;
  datasetId: number;
  name: string;
  filePath: string;
  fileType: string;
  totalRows: number;
  columns: TableColumn[];
  rows: TableRow[];
  importDate?: string;
  executionState: string;
  errorMessage?: string;
}

export interface GraphNodePreview {
  id: string;
  label?: string;
  layer?: string;
  weight?: number;
  isPartition: boolean;
  attrs?: Record<string, any>;
}

export interface GraphEdgePreview {
  id: string;
  source: string;
  target: string;
  label?: string;
  layer?: string;
  weight?: number;
  attrs?: Record<string, any>;
}

export interface GraphPreview {
  nodeId: string;
  graphId: number;
  name: string;
  annotations?: string | null;
  nodes: GraphNodePreview[];
  edges: GraphEdgePreview[];
  layers: Layer[];
  nodeCount: number;
  edgeCount: number;
  executionState: string;
  computedDate?: string;
  errorMessage?: string;
}

export interface GetDataSetPreviewResponse {
  datasetPreview: DataSetPreview | null;
}

export interface GetGraphPreviewResponse {
  graphPreview: GraphPreview | null;
}

// Query Variables

export interface GetDataSetPreviewVariables {
  projectId: number;
  nodeId: string;
  limit?: number;
  offset?: number;
}

export interface GetGraphPreviewVariables {
  projectId: number;
  nodeId: string;
}

// Execution States
export enum ExecutionState {
  NOT_STARTED = 'not_started',
  PENDING = 'pending',
  PROCESSING = 'processing',
  COMPLETED = 'completed',
  ERROR = 'error',
}

// Helper to check if execution is complete
export function isExecutionComplete(state: string): boolean {
  return state === ExecutionState.COMPLETED;
}

// Helper to check if execution failed
export function isExecutionFailed(state: string): boolean {
  return state === ExecutionState.ERROR;
}

// Helper to check if execution is in progress
export function isExecutionInProgress(state: string): boolean {
  return state === ExecutionState.PENDING || state === ExecutionState.PROCESSING;
}

// Helper to get execution state display label
export function getExecutionStateLabel(state: string): string {
  switch (state) {
    case ExecutionState.NOT_STARTED:
      return 'Not Started';
    case ExecutionState.PENDING:
      return 'Pending';
    case ExecutionState.PROCESSING:
      return 'Processing';
    case ExecutionState.COMPLETED:
      return 'Ready';
    case ExecutionState.ERROR:
      return 'Error';
    default:
      return state;
  }
}

// Execute Node Mutation
export const EXECUTE_NODE = gql`
  mutation ExecuteNode($projectId: Int!, $nodeId: String!) {
    executeNode(projectId: $projectId, nodeId: $nodeId) {
      success
      message
      nodeId
    }
  }
`;

export interface NodeExecutionResult {
  success: boolean;
  message: string;
  nodeId: string;
}

// Execute Plan (DAG) Mutation
export const EXECUTE_PLAN = gql`
  mutation ExecutePlan($projectId: Int!, $planId: Int!) {
    executePlan(projectId: $projectId, planId: $planId) {
      success
      message
      outputFiles
    }
  }
`;

export interface PlanExecutionResult {
  success: boolean;
  message: string;
  outputFiles: string[];
}

// Clear Project Execution State Mutation (resets all graph data, keeps config and datasets)
export const CLEAR_PROJECT_EXECUTION = gql`
  mutation ClearProjectExecution($projectId: Int!) {
    clearProjectExecution(projectId: $projectId) {
      success
      message
    }
  }
`;

// Stop Plan Execution Mutation
export const STOP_PLAN_EXECUTION = gql`
  mutation StopPlanExecution($projectId: Int!) {
    stopPlanExecution(projectId: $projectId) {
      success
      message
    }
  }
`;

export interface ExecutionActionResult {
  success: boolean;
  message: string;
}

// Helper to get execution state color
export function getExecutionStateColor(state: string): string {
  switch (state) {
    case ExecutionState.NOT_STARTED:
      return 'gray';
    case ExecutionState.PENDING:
      return 'yellow';
    case ExecutionState.PROCESSING:
      return 'blue';
    case ExecutionState.COMPLETED:
      return 'green';
    case ExecutionState.ERROR:
      return 'red';
    default:
      return 'gray';
  }
}

File: layercake-core/src/graphql/types/preview.rs
use async_graphql::*;
use serde::{Deserialize, Serialize};

/// DataSet preview with table data
#[derive(Clone, Debug, SimpleObject)]
pub struct DataSetPreview {
    pub node_id: String,
    pub dataset_id: i32,
    pub name: String,
    pub file_path: String,
    pub file_type: String,
    pub total_rows: i32,
    pub columns: Vec<TableColumn>,
    pub rows: Vec<TableRow>,
    pub import_date: Option<String>,
    pub execution_state: String,
    pub error_message: Option<String>,
}

/// Column metadata for table preview
#[derive(Clone, Debug, SimpleObject, Serialize, Deserialize)]
pub struct TableColumn {
    pub name: String,
    pub data_type: String,
    pub nullable: bool,
}

/// Row data for table preview
#[derive(Clone, Debug, SimpleObject)]
pub struct TableRow {
    pub row_number: i32,
    pub data: serde_json::Value,
}

/// Graph preview with nodes and edges
#[derive(Clone, Debug, SimpleObject)]
pub struct GraphPreview {
    pub node_id: String,
    pub graph_id: i32,
    pub name: String,
    pub annotations: Option<String>,
    pub nodes: Vec<GraphNodePreview>,
    pub edges: Vec<GraphEdgePreview>,
    pub layers: Vec<crate::graphql::types::layer::Layer>,
    pub node_count: i32,
    pub edge_count: i32,
    pub execution_state: String,
    pub computed_date: Option<String>,
    pub error_message: Option<String>,
}

/// Graph node for preview
#[derive(Clone, Debug, SimpleObject)]
pub struct GraphNodePreview {
    pub id: String,
    pub label: Option<String>,
    pub layer: Option<String>,
    pub weight: Option<f64>,
    pub is_partition: bool,
    /// Deprecated: use attributes
    pub attrs: Option<serde_json::Value>,
    pub attributes: Option<serde_json::Value>,
}

/// Graph edge for preview
#[derive(Clone, Debug, SimpleObject)]
pub struct GraphEdgePreview {
    pub id: String,
    pub source: String,
    pub target: String,
    pub label: Option<String>,
    pub layer: Option<String>,
    pub weight: Option<f64>,
    /// Deprecated: use attributes
    pub attrs: Option<serde_json::Value>,
    pub attributes: Option<serde_json::Value>,
}

impl From<crate::database::entities::graph_nodes::Model> for GraphNodePreview {
    fn from(model: crate::database::entities::graph_nodes::Model) -> Self {
        Self {
            id: model.id,
            label: model.label,
            layer: model.layer,
            weight: model.weight,
            is_partition: model.is_partition,
            attrs: model.attrs.clone(),
            attributes: model.attrs,
        }
    }
}

impl From<crate::database::entities::graph_data_nodes::Model> for GraphNodePreview {
    fn from(model: crate::database::entities::graph_data_nodes::Model) -> Self {
        Self {
            id: model.external_id,
            label: model.label,
            layer: model.layer,
            weight: model.weight,
            is_partition: model.is_partition,
            attrs: model.attributes.clone(),
            attributes: model.attributes,
        }
    }
}

impl From<crate::database::entities::graph_edges::Model> for GraphEdgePreview {
    fn from(model: crate::database::entities::graph_edges::Model) -> Self {
        Self {
            id: model.id,
            source: model.source,
            target: model.target,
            label: model.label,
            layer: model.layer,
            weight: model.weight,
            attrs: model.attrs.clone(),
            attributes: model.attrs,
        }
    }
}

impl From<crate::database::entities::graph_data_edges::Model> for GraphEdgePreview {
    fn from(model: crate::database::entities::graph_data_edges::Model) -> Self {
        Self {
            id: model.external_id,
            source: model.source,
            target: model.target,
            label: model.label,
            layer: model.layer,
            weight: model.weight,
            attrs: model.attributes.clone(),
            attributes: model.attributes,
        }
    }
}
